graphql_compiler/compiler/cypher_query.py

```diff
@@ -0,0 +1,168 @@
+    """Return a CypherStep for a list of IR blocks that start with a QueryRoot block."""
+    current_step_block_types = tuple(type(block) for block in current_step_blocks)
+
+    if current_step_block_types == (QueryRoot, Filter, MarkLocation):
+        step_block, where_block, as_block = current_step_blocks
+    elif current_step_block_types == (QueryRoot, MarkLocation):
+        where_block = None
+        step_block, as_block = current_step_blocks
+    else:
+        raise AssertionError(u'Unexpected current_step_blocks received: {}'
+                             .format(current_step_blocks))
+
+    exact_step_type = get_only_element_from_collection(step_block.start_class)
+    step_types = _get_all_supertypes_of_exact_type(query_metadata_table, exact_step_type)
+
+    return CypherStep(
+        linked_location=linked_location, step_block=step_block, step_types=step_types,
+        where_block=where_block, as_block=as_block)
+
+
+##############
+# Public API #
+##############
+
+def convert_to_cypher_query(ir_blocks, query_metadata_table, type_equivalence_hints=None):
+    """Convert the list of IR blocks into a CypherQuery object, for easier manipulation."""
+    steps = []
+    current_step_blocks = None
+    linked_location = None
+    next_linked_location = None
+
+    folds, remaining_ir_blocks = extract_folds_from_ir_blocks(ir_blocks)
+
+    global_operations_index = None
+
+    for current_block_index, block in enumerate(remaining_ir_blocks):
+        if isinstance(block, QueryRoot):
+            if current_step_blocks is not None:
+                raise AssertionError(u'Unexpectedly encountered a QueryRoot block that was not '
+                                     u'the first block in the IR: {} {}'
+                                     .format(block, remaining_ir_blocks))
+
+            current_step_blocks = [block]
+        elif isinstance(block, step_block_types):
+            cypher_step = _make_cypher_step(
+                query_metadata_table, linked_location, current_step_blocks)
+            steps.append(cypher_step)
+
+            linked_location = next_linked_location
+            next_linked_location = None
+
+            current_step_blocks = [block]
+        elif isinstance(block, GlobalOperationsStart):
+            global_operations_index = current_block_index
+            break
+        elif isinstance(block, linked_location_source_types):
+            next_linked_location = block.location
+            if isinstance(block, MarkLocation):
+                current_step_blocks.append(block)
+        elif isinstance(block, discarded_block_types):
+            pass
+        elif isinstance(block, (Filter, CoerceType)):
+            current_step_blocks.append(block)
+        else:
+            raise AssertionError(u'Unexpected block encountered: {} {}'
+                                 .format(block, ir_blocks))
+
+    steps.append(_make_cypher_step(query_metadata_table, linked_location, current_step_blocks))
```

Comment:
It seems like every query has a GlobalOperationsStart block, which means that the only way to break out of the for-loop above is by coming across that block. 

Would it make more sense to put this line of code in the GlobalOperationsStart case above instead of as a standalone line of code? This makes it clearer that we're adding one last cypher step.

Reply:
Good idea!